Maple2.File.IO/Nif/NifDocument.cs

```diff
@@ -0,0 +1,225 @@
+﻿using System.IO;
+using System.Text;
+
+namespace Maple2.File.IO.Nif;
+
+public class NifDocument {
+    private readonly byte[] fileData;
+    private NifDocumentHeader header;
+    public EndianReader Reader { get; private set; }
+    public string RelPath { get; init; }
+    private NifBlock[] blocks;
+    public NifBlock? ReadingBlock { get; private set; }
+
+    public List<NiPhysXProp> PhysXProps { get; init; }
+    public string VersionString { get => header.HeaderString; }
+
+    public NifDocument(string relpath, byte[] fileData) {
+        PhysXProps = new List<NiPhysXProp>();
+        this.fileData = fileData;
+
+        Reader = new EndianReader(new MemoryStream(), false);
+        blocks = Array.Empty<NifBlock>();
+        header = new NifDocumentHeader();
+        RelPath = relpath;
+    }
+
+    private int ReadHeaderString() {
+        int headerStringLength = Array.IndexOf(fileData, (byte) '\n');
+
+        header.HeaderString = Encoding.UTF8.GetString(fileData, 0, headerStringLength);
+
+        int versionStart = header.HeaderString.LastIndexOf(' ');
+        string versionNumbers = header.HeaderString.Substring(versionStart);
+
+        header.Version = Array.ConvertAll(versionNumbers.Split('.'), ushort.Parse);
+
+        return headerStringLength + 1;
+    }
+
+    public NifBlock? GetBlock(int index) {
+        if (index == -1) {
+            return null;
+        }
+
+        if (blocks[index] is not null) {
+            return blocks[index];
+        }
+
+        string blockType = header.BlockTypes[header.BlockTypeIndices[index]];
+
+        NifBlock block = blockType switch {
+            "NiPhysXProp" => new NiPhysXProp(index),
+            "NiPhysXPropDesc" => new NiPhysXPropDesc(index),
+            "NiPhysXActorDesc" => new NiPhysXActorDesc(index),
+            "NiPhysXShapeDesc" => new NiPhysXShapeDesc(index),
+            "NiPhysXMeshDesc" => new NiPhysXMeshDesc(index),
+            _ => new NifBlock(blockType, false, index)
+        };
+
+        blocks[index] = block;
+
+        return block;
+    }
+
+    public T? GetBlock<T>(int index) {
+        if (index == -1) {
```

Comment:
Here and below: do we actually need to allow -1? If we didn't the return value wouldn't need to be nullable. If not, we can just throw [ArgumentOutOfRangeException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception?view=net-8.0)

Reply:
It's required in `ReadBlockRef` because the NIF standard uses it for null values, but looks like GetBlock can be made not nullable since that's already wrapped by ReadBlockRef.
```
System.InvalidOperationException: Error reading nif: [/library/standardmodellibrary/media/camera.nif] (Gamebryo File Format, Version 30.0.0.2) in block [13] NiPhysXProp "PhysX Prop"
 ---> System.Exception: ReadBlockRef(-1)
```